diff -Nur hc-cron-0.14/crontab.c hc-cron-0.14-new/crontab.c
--- hc-cron-0.14/crontab.c	Sat Mar 10 16:20:48 2001
+++ hc-cron-0.14-new/crontab.c	Mon Mar  4 13:15:32 2002
@@ -405,6 +405,8 @@
 		goto fatal;
 	}
 	mtime = statbuf.st_mtime;
+	fclose(NewCrontab);
+	close(t);
 
 	if ((!(editor = getenv("VISUAL")))
 	 && (!(editor = getenv("EDITOR")))
@@ -468,6 +470,15 @@
 			WCOREDUMP(waiter) ?"" :"no ");
 		goto fatal;
 	}
+	if (-1 == (t = open(Filename, O_RDWR))) {
+		perror(Filename);
+		goto fatal;
+	}
+	if (!(NewCrontab = fdopen(t, "r+"))) {
+		perror("fdopen");
+		goto fatal;
+	}
+
 	if (fstat(t, &statbuf) < 0) {
 		perror("fstat");
 		goto fatal;
