--- hc-cron-0.13/man/crontab.1.wiget	Sun Sep 26 01:39:41 1999
+++ hc-cron-0.13/man/crontab.1	Mon Oct 11 15:36:40 1999
@@ -93,8 +93,8 @@
 crontab(5), cron(8)
 .SH FILES
 .nf
-/etc/cron.allow
-/etc/cron.deny
+/etc/cron/cron.allow
+/etc/cron/cron.deny
 .fi
 .SH STANDARDS
 The
--- hc-cron-0.13/Makefile.wiget	Sun Sep 26 01:39:41 1999
+++ hc-cron-0.13/Makefile	Mon Oct 11 15:39:26 1999
@@ -46,7 +46,7 @@
 DESTROOT	=	$(DESTDIR)/usr
 DESTSBIN	=	$(DESTROOT)/sbin
 DESTBIN		=	$(DESTROOT)/bin
-DESTMAN		=	$(DESTROOT)/man
+DESTMAN		=	$(DESTROOT)/share/man
 #<<need bitstring.h>>
 INCLUDE		=	-I.
 #<<need getopt()>>
@@ -96,13 +96,11 @@
 			$(CC) $(LDFLAGS) -o crontab $(CRONTAB_OBJ) $(LIBS)
 
 install		:	all
-			$(INSTALL) -c -m 750 -o root -s cron    $(DESTSBIN)/crond
-			$(INSTALL) -c -m 4755 -o root -s crontab $(DESTBIN)/
-			$(INSTALL) -c -m 644 -o root man/crontab.1 $(DESTMAN)/man1/
-			$(INSTALL) -c -m 644 -o root man/crontab.5 $(DESTMAN)/man5/
-			$(INSTALL) -c -m 644 -o root man/cron.8 $(DESTMAN)/man8/
-			rm -f $(DESTMAN)/man8/crond.8
-			ln -sf $(DESTMAN)/man8/cron.8 $(DESTMAN)/man8/crond.8
+			$(INSTALL) -c -m 750  -s cron    $(DESTSBIN)/crond
+			$(INSTALL) -c -m 4755 -s crontab $(DESTBIN)/
+			$(INSTALL) -c -m 644 man/crontab.1 $(DESTMAN)/man1/
+			$(INSTALL) -c -m 644 man/crontab.5 $(DESTMAN)/man5/
+			$(INSTALL) -c -m 644 man/cron.8 $(DESTMAN)/man8/
 clean		:
 			rm -f *.o cron crontab a.out core tags *~ #*
 
--- hc-cron-0.13/pathnames.h.wiget	Mon Oct 11 15:36:40 1999
+++ hc-cron-0.13/pathnames.h	Mon Oct 11 15:36:40 1999
@@ -47,15 +47,15 @@
 			 * LOG_FILE or SYSLOG is defined, we don't log.  If
 			 * both are defined, we log both ways.
 			 */
-#define	ALLOW_FILE	"/etc/cron.allow"		/*-*/
-#define DENY_FILE	"/etc/cron.deny"		/*-*/
+#define	ALLOW_FILE	"/etc/cron/cron.allow"		/*-*/
+#define DENY_FILE	"/etc/cron/cron.deny"		/*-*/
 #define LOG_FILE	"/var/log/cron"		/*-*/
 
 			/* this file keeps track of when cron was last run so
 			 * that cron can check for jobs it should have started
 			 * during downtime. Added by hcl feb98
 			 */
-#define LASTRUN_FILE	"/var/lib/cron.lastrun"
+#define LASTRUN_FILE	"/var/lib/misc/cron.lastrun"
 
 			/* we get the system loadaverage from here.
 			 * If not defined RUN_ONLY_IDLE is disabled.
@@ -82,7 +82,7 @@
 #if defined(_PATH_VI)
 # define EDITOR _PATH_VI
 #else
-# define EDITOR "/usr/ucb/vi"
+# define EDITOR "/bin/vi"
 #endif
 
 #ifndef _PATH_BSHELL
